src/Sylius/Component/Core/Inventory/Operator/OrderInventoryOperator.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Inventory\Operator;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class OrderInventoryOperator implements OrderInventoryOperatorInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function updateInventoryState(OrderInterface $order) {
+        /** @var OrderItemInterface $orderItem */
+        foreach ($order->getItems() as $orderItem) {
+            $variant = $orderItem->getVariant();
+
+            if (!$variant->isTracked()) {
+                continue;
+            }
+
+            if ('paid' === $order->getPaymentState()) {
+
+                if ('cancelled' === $order->getState()) {
+                    $variant->setOnHand($variant->getOnHand() + $orderItem->getQuantity());
+                    continue;
+                }
+
+                Assert::greaterThanEq(
+                    ($variant->getOnHold() - $orderItem->getQuantity()),
+                    0,
+                    sprintf('Not enough units to decrease on hold quantity from the inventory of a variant "%s".', $variant->getName())
+                );
+
+                Assert::greaterThanEq(
+                    ($variant->getOnHand() - $orderItem->getQuantity()),
+                    0,
+                    sprintf('Not enough units to decrease on hand quantity from the inventory of a variant "%s".', $variant->getName())
+                );
+
+                $variant->setOnHand($variant->getOnHand() - $orderItem->getQuantity());
+                $variant->setOnHold($variant->getOnHold() - $orderItem->getQuantity());
+                continue;
+            }
+
+            if ('cancelled' === $order->getState() && 'paid' !== $order->getPaymentState()) {
```

Comment:
It seems that the 2nd check will always be fulfilled as you check for the opposite a couple of lines up.